plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableChangesFunction.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.trino.plugin.iceberg.catalog.TrinoCatalogFactory;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorTransactionHandle;
+import io.trino.spi.connector.SchemaTableName;
+import io.trino.spi.ptf.AbstractConnectorTableFunction;
+import io.trino.spi.ptf.Argument;
+import io.trino.spi.ptf.Descriptor;
+import io.trino.spi.ptf.ScalarArgument;
+import io.trino.spi.ptf.ScalarArgumentSpecification;
+import io.trino.spi.ptf.TableFunctionAnalysis;
+import io.trino.spi.type.TypeManager;
+import io.trino.spi.type.VarcharType;
+import org.apache.iceberg.Table;
+
+import javax.inject.Inject;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.plugin.iceberg.TypeConverter.toTrinoType;
+import static io.trino.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.trino.spi.ptf.ReturnTypeSpecification.GenericTable.GENERIC_TABLE;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_MILLIS;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergTableChangesFunction
+        extends AbstractConnectorTableFunction
+{
+    private static final String START_SNAPSHOT_VAR_NAME = "START_SNAPSHOT_ID";
+    private static final String END_SNAPSHOT_VAR_NAME = "END_SNAPSHOT_ID";
+
+    private final TrinoCatalogFactory trinoCatalogFactory;
+    private final TypeManager typeManager;
+
+    @Inject
+    public IcebergTableChangesFunction(TrinoCatalogFactory trinoCatalogFactory, TypeManager typeManager)
```

Comment:
This table function throws an exception if the table definition has changed. 
```
trino:tpch> select * from table ...;
 c1 | _change_type | _change_version_id  |      _change_timestamp
----+--------------+---------------------+-----------------------------
  1 | insert       | 6985197716516736310 | 2023-01-12 10:02:11.362 UTC

trino:tpch> alter table test2 add column c2 int;
ADD COLUMN

trino:tpch> select * from table ...;
returned table does not match the node's output
java.lang.IllegalStateException: returned table does not match the node's output
	at com.google.common.base.Preconditions.checkState(Preconditions.java:502)
	at io.trino.sql.planner.iterative.rule.RewriteTableFunctionToTableScan.apply(RewriteTableFunctionToTableScan.java:67)
	at io.trino.sql.planner.iterative.rule.RewriteTableFunctionToTableScan.apply(RewriteTableFunctionToTableScan.java:38)
	at io.trino.sql.planner.iterative.IterativeOptimizer.transform(IterativeOptimizer.java:203)
	at io.trino.sql.planner.iterative.IterativeOptimizer.exploreNode(IterativeOptimizer.java:170)
	at io.trino.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:133)
	at io.trino.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:252)
	at io.trino.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:135)
	at io.trino.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:118)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:247)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:222)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:217)
	at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:475)
	at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:456)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:397)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:249)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_testversion____20230112_095355_71.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Reply:
Thanks, I need to think a bit more about which schema should be used for these situations. The schema of the first snapshot in the range, last snapshot in the range?

Reply:
last one